File: cloudflare/src/privacy-pass.js
/**
 * Privacy Pass at the edge (RFC 9576/9577/9578) with @cloudflare/privacypass-ts.
 *
 * The Worker is both Issuer and Origin. Publicly verifiable tokens (type
 * 0x0002, Blind RSA, PSS mode): one token pays for one priced API call.
 *
 *   GET  /.well-known/private-token-issuer-directory   (RFC 9578 §4.3)
 *   POST /token-request                                (single or generic batch, by Content-Type)
 *   POST /token-request/batch/{n}                      (generic batch, fixed-price for x402)
 *   Authorization: PrivateToken token=...              on /v1/* priced routes
 *
 * Secrets/vars: PRIVACY_PASS_PRIVATE_JWK (issuing key), PRIVACY_PASS_OLD_PUBLIC_JWKS
 * (JSON array; still redeemable), binding PRIVACY_PASS_DB (D1 nullifier ledger).
 */

import { Hono } from "hono";
import { publicVerif, genericBatched, AuthorizationHeader, WWWAuthenticateHeader, TOKEN_TYPES, MediaType, PRIVATE_TOKEN_ISSUER_DIRECTORY } from "@cloudflare/privacypass-ts";

const MODE = publicVerif.BlindRSAMode.PSS; // RFC 9578 §6: PSS with 48-byte salt
const ALGO = { name: "RSA-PSS", hash: "SHA-384" };
const hex = (u8) => [...u8].map((b) => b.toString(16).padStart(2, "0")).join("");

const pools = new WeakMap();
async function pool(env) {
  let p = pools.get(env);
  if (p) return p;
  if (!env.PRIVACY_PASS_PRIVATE_JWK) return null;
  const priv = JSON.parse(env.PRIVACY_PASS_PRIVATE_JWK);
  const privateKey = await crypto.subtle.importKey("jwk", { ...priv, alg: "PS384", ext: true }, ALGO, true, ["sign"]);
  const pubJwk = { kty: "RSA", n: priv.n, e: priv.e, alg: "PS384", ext: true };
  const publicKey = await crypto.subtle.importKey("jwk", pubJwk, ALGO, true, ["verify"]);
  const issuerName = (env.PRIVACY_PASS_ISSUER_NAME || "").trim() || null;
  const issuer = new publicVerif.Issuer(MODE, issuerName || "issuer", privateKey, publicKey);
  const issuing = { publicKey, spki: await publicVerif.getPublicKeyBytes(publicKey), keyId: await issuer.tokenKeyID(), issuing: true };
  const keys = [issuing];
  for (const old of JSON.parse(env.PRIVACY_PASS_OLD_PUBLIC_JWKS || "[]")) {
    const pk = await crypto.subtle.importKey("jwk", { kty: "RSA", n: old.n, e: old.e, alg: "PS384", ext: true }, ALGO, true, ["verify"]);
    const spki = await publicVerif.getPublicKeyBytes(pk);
    keys.push({ publicKey: pk, spki, keyId: new Uint8Array(await crypto.subtle.digest("SHA-256", spki)), issuing: false });
  }
  p = { issuer, batched: new genericBatched.Issuer(issuer), issuing, keys, issuerName };
  pools.set(env, p);
  return p;
}

/** Origin role: challenge + redemption against the D1 ledger. */
export class PrivacyPassOrigin {
  static async get(env, url) {
    const p = await pool(env);
    if (!p) return null;
    const host = (env.PRIVACY_PASS_ORIGIN || "").trim() || new URL(url).host;
    return new PrivacyPassOrigin(env, p, host);
  }
  constructor(env, p, host) {
    this.env = env; this.pool = p; this.host = host;
    this.origin = new publicVerif.Origin(MODE, [host]);
    // Static per-origin challenge (empty redemption context): tokens are pre-purchasable.
    this.challenge = this.origin.createTokenChallenge(p.issuerName || host, new Uint8Array(0));
  }
  challengeHeader() {
    return new WWWAuthenticateHeader(this.challenge, this.pool.issuing.spki, 86400 * 30).toString();
  }
  async redeem(authorization) {
    let headers;
    try { headers = AuthorizationHeader.parse(TOKEN_TYPES.BLIND_RSA, authorization); } catch { return { ok: false, why: "malformed PrivateToken header" }; }
    if (!headers.length) return { ok: false, why: "no PrivateToken credential" };
    const token = headers[0].token;
    const key = this.pool.keys.find((k) => hex(k.keyId) === hex(token.authInput.tokenKeyId));
    if (!key) return { ok: false, why: "unknown token key" };
    const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", this.challenge.serialize()));
    if (hex(digest) !== hex(token.authInput.challengeDigest)) return { ok: false, why: "token was issued for a different challenge" };
    if (!(await this.origin.verify(token, key.publicKey))) return { ok: false, why: "invalid token" };
    if (!this.env.PRIVACY_PASS_DB) return { ok: false, why: "ledger not configured" };
    const nullifier = hex(new Uint8Array(await crypto.subtle.digest("SHA-256", new Uint8Array([...new TextEncoder().encode("nullifier"), ...token.authInput.nonce]))));
    const r = await this.env.PRIVACY_PASS_DB.prepare("INSERT OR IGNORE INTO spent (nullifier, kid, ts) VALUES (?1, ?2, ?3)").bind(nullifier, hex(key.keyId).slice(0, 12), Date.now()).run();
    if (!Number(r.meta?.changes ?? 0)) return { ok: false, why: "token already spent" };
    await this.env.PRIVACY_PASS_DB.prepare("INSERT INTO stats (kid, issued, redeemed) VALUES (?1, 0, 1) ON CONFLICT(kid) DO UPDATE SET redeemed = redeemed + 1").bind(hex(this.pool.issuing.keyId).slice(0, 12)).run();
    return { ok: true };
  }
}

async function countIssued(env, kid, n) {
  if (!env.PRIVACY_PASS_DB) return;
  await env.PRIVACY_PASS_DB.prepare("INSERT INTO stats (kid, issued, redeemed) VALUES (?1, ?2, 0) ON CONFLICT(kid) DO UPDATE SET issued = issued + ?2").bind(kid, n).run();
}

export function privacyPassRoutes({ VERSION }) {
  const r = new Hono();
  const b64 = (u8) => btoa(String.fromCharCode(...u8)); // padded base64url per RFC 9577
  const b64url = (u8) => b64(u8).replace(/\+/g, "-").replace(/\//g, "_");

  r.get(PRIVATE_TOKEN_ISSUER_DIRECTORY, async (c) => {
    const p = await pool(c.env);
    if (!p) return c.json({ detail: "Privacy Pass issuer not configured (set PRIVACY_PASS_PRIVATE_JWK)." }, 404);
    const body = { "issuer-request-uri": "/token-request", "token-keys": p.keys.map((k) => ({ "token-type": TOKEN_TYPES.BLIND_RSA.value, "token-key": b64url(k.spki), ...(k.issuing ? {} : { "not-after": null }) })) };
    return new Response(JSON.stringify(body), { headers: { "content-type": MediaType.PRIVATE_TOKEN_ISSUER_DIRECTORY, "cache-control": "public, max-age=300" } });
  });

  // A resource that always challenges: the discoverable way to fetch the challenge
  // (RFC 9577 puts it on 401/402 responses; the free edge API rarely returns those).
  r.get("/token-request/challenge", async (c) => {
    const origin = await PrivacyPassOrigin.get(c.env, c.req.url);
    if (!origin) return c.json({ detail: "Privacy Pass issuer not configured." }, 404);
    return c.json({ detail: "Present a PrivateToken (see WWW-Authenticate) or buy tokens at POST /token-request." }, 401, { "www-authenticate": origin.challengeHeader() });
  });

  r.get("/token-request/stats", async (c) => {
    const p = await pool(c.env);
    if (!p) return c.json({ detail: "not configured" }, 404);
    let stats = { issued: 0, redeemed: 0 };
    if (c.env.PRIVACY_PASS_DB) { const row = await c.env.PRIVACY_PASS_DB.prepare("SELECT COALESCE(SUM(issued),0) AS issued, COALESCE(SUM(redeemed),0) AS redeemed FROM stats").first(); if (row) stats = { issued: Number(row.issued), redeemed: Number(row.redeemed) }; }
    return c.json({ standard: "RFC 9578 type 0x0002 (Blind RSA, PSS)", keys: p.keys.map((k) => ({ kid: hex(k.keyId).slice(0, 12), issuing: k.issuing })), pool: { ...stats, outstanding: stats.issued - stats.redeemed }, version: VERSION });
  });

  const single = async (c, p, bytes) => {
    let req;
    try { req = publicVerif.TokenRequest.deserialize(TOKEN_TYPES.BLIND_RSA, bytes); } catch { return c.json({ detail: "malformed TokenRequest" }, 422); }
    if (req.truncatedTokenKeyId !== p.issuing.keyId[p.issuing.keyId.length - 1]) return c.json({ detail: "TokenRequest is for a key this issuer does not issue with" }, 422);
    let res;
    try { res = await p.issuer.issue(req); } catch (e) { return c.json({ detail: `issuance failed: ${e.message}` }, 422); }
    await countIssued(c.env, hex(p.issuing.keyId).slice(0, 12), 1);
    return new Response(res.serialize(), { headers: { "content-type": MediaType.PRIVATE_TOKEN_RESPONSE, "cache-control": "no-store" } });
  };
  const batch = async (c, p, bytes, expected) => {
    let reqs;
    try { reqs = genericBatched.BatchedTokenRequest.deserialize(bytes); } catch { return c.json({ detail: "malformed BatchedTokenRequest" }, 422); }
    const n = reqs.tokenRequests.length;
    if (expected !== null && n !== expected) return c.json({ detail: `this batch endpoint issues exactly ${expected} tokens; ${n} requested` }, 422);
    if (n < 1 || n > 100) return c.json({ detail: "batches are 1-100 tokens" }, 422);
    const res = await p.batched.issue(reqs);
    const issued = res.tokenResponses.filter((t) => t.tokenResponse !== null).length;
    await countIssued(c.env, hex(p.issuing.keyId).slice(0, 12), issued);
    return new Response(res.serialize(), { headers: { "content-type": MediaType.GENERIC_BATCHED_TOKEN_RESPONSE, "cache-control": "no-store", "x-tokens-issued": String(issued) } });
  };

  r.post("/token-request", async (c) => {
    const p = await pool(c.env);
    if (!p) return c.json({ detail: "Privacy Pass issuer not configured." }, 404);
    const ct = (c.req.header("content-type") || "").split(";")[0].trim();
    const bytes = new Uint8Array(await c.req.arrayBuffer());
    if (ct === MediaType.GENERIC_BATCHED_TOKEN_REQUEST) return batch(c, p, bytes, null);
    if (ct !== MediaType.PRIVATE_TOKEN_REQUEST) return c.json({ detail: `Content-Type must be ${MediaType.PRIVATE_TOKEN_REQUEST} or ${MediaType.GENERIC_BATCHED_TOKEN_REQUEST}` }, 415);
    return single(c, p, bytes);
  });
  r.post("/token-request/batch/:n", async (c) => {
    const p = await pool(c.env);
    if (!p) return c.json({ detail: "Privacy Pass issuer not configured." }, 404);
    const n = Number(c.req.param("n"));
    if (![10, 25, 100].includes(n)) return c.json({ detail: "batch denominations: 10, 25, 100" }, 404);
    return batch(c, p, new Uint8Array(await c.req.arrayBuffer()), n);
  });
  return r;
}

File: timesfm3/serving/privacypass.py
"""Server side of Privacy Pass for the service: issuer, origin, ledger, keys.

One token pays for one priced call. Keys are private JWKs (with CRT
parameters) shared with the Cloudflare Worker, so tokens issued by either
redeem at either. Rotation is lossless: the issuing key signs new tokens,
older keys stay redeemable.

    TIMESFM3_PRIVACY_PASS_KEY_FILE=keys/pp-2026-09.json     # issuing (created if missing)
    TIMESFM3_PRIVACY_PASS_OLD_KEYS=keys/pp-2026-06.json,... # redeem-only
    TIMESFM3_PRIVACY_PASS_LEDGER_FILE=pp-ledger.json         # spent nonces, counters
    TIMESFM3_PRIVACY_PASS_ISSUER_NAME=api.example.com        # default: request host
"""

from __future__ import annotations

import hashlib
import json
import math
import os
import secrets
import tempfile
import threading

from .. import blindrsa as B
from .. import privacypass as PP
from .auth import ApiKey

TOKEN_IDENTITY = ApiKey(key="", name="privacy-pass", plan="prepaid")
DENOMINATIONS = (10, 25, 100)
DEFAULT_PRICE_PER_TOKEN_USD = 0.004
POINTS_PER_TOKEN = 256  # what a plan holder is charged per token bought


# -- RSA key generation in pure Python (runs once; persisted as a JWK).

_SMALL_PRIMES = [p for p in range(3, 2000, 2) if all(p % q for q in range(3, int(p ** 0.5) + 1, 2))]


def _is_probable_prime(n: int, rounds: int = 40) -> bool:
    if n < 2:
        return False
    for p in _SMALL_PRIMES:
        if n % p == 0:
            return n == p
    d, r = n - 1, 0
    while d % 2 == 0:
        d //= 2
        r += 1
    for _ in range(rounds):
        a = secrets.randbelow(n - 3) + 2
        x = pow(a, d, n)
        if x in (1, n - 1):
            continue
        for _ in range(r - 1):
            x = pow(x, 2, n)
            if x == n - 1:
                break
        else:
            return False
    return True


def _random_prime(bits: int) -> int:
    while True:
        c = secrets.randbits(bits) | (1 << (bits - 1)) | 1
        if _is_probable_prime(c):
            return c


def generate_private_jwk(bits: int = 2048, e: int = 65537) -> dict:
    """A private RSA JWK (with CRT parameters) usable by Python and WebCrypto."""
    while True:
        p, q = _random_prime(bits // 2), _random_prime(bits // 2)
        if p == q:
            continue
        n = p * q
        phi = (p - 1) * (q - 1)
        if n.bit_length() == bits and math.gcd(e, phi) == 1:
            break
    if p < q:
        p, q = q, p
    return B.private_jwk(B.PrivateKey(n, e, pow(e, -1, phi)), p, q)


class PoolKey:
    def __init__(self, jwk: dict, issuing: bool):
        self.private = B.private_from_jwk(jwk) if "d" in jwk else None
        self.public = B.public_from_jwk(jwk)
        self.key_id = PP.token_key_id(self.public)
        self.kid = self.key_id.hex()[:12]
        self.issuing = issuing


class PrivacyPassService:
    """Issuer + Origin + double-spend ledger for the FastAPI service."""

    def __init__(self, key_file: str | None = None, ledger_file: str | None = None,
                 price_per_token_usd: float = DEFAULT_PRICE_PER_TOKEN_USD, bits: int = 2048,
                 old_key_files: list[str] | None = None, issuer_name: str | None = None,
                 origin_name: str | None = None):
        self._lock = threading.Lock()
        self.ledger_file = ledger_file
        self.price_per_token_usd = price_per_token_usd
        self.issuer_name = issuer_name
        self.origin_name = origin_name  # default: the request host (per-origin tokens)
        self.ephemeral = key_file is None
        if key_file and os.path.exists(key_file):
            with open(key_file) as f:
                jwk = json.load(f)
            if "kty" not in jwk:
                raise ValueError(f"{key_file} is not a JWK; regenerate it with scripts/credits_keygen.py")
        else:
            jwk = generate_private_jwk(bits)
            if key_file:
                d = os.path.dirname(os.path.abspath(key_file))
                os.makedirs(d, exist_ok=True)
                fd, tmp = tempfile.mkstemp(dir=d, prefix=".pp-key-")
                with os.fdopen(fd, "w") as f:
                    json.dump(jwk, f)
                os.chmod(tmp, 0o600)
                os.replace(tmp, key_file)
        self.issuing_key = PoolKey(jwk, issuing=True)
        self.keys: dict[bytes, PoolKey] = {self.issuing_key.key_id: self.issuing_key}
        for path in old_key_files or []:
            with open(path) as f:
                k = PoolKey(json.load(f), issuing=False)
            self.keys.setdefault(k.key_id, k)
        self.issuer = PP.Issuer([self.issuing_key.private])
        self.origin = PP.Origin([k.public for k in self.keys.values()])
        self.spent: set[str] = set()
        self.issued = 0
        self.redeemed = 0
        if ledger_file and os.path.exists(ledger_file):
            with open(ledger_file) as f:
                doc = json.load(f)
            self.spent = set(doc.get("spent", []))
            self.issued = int(doc.get("issued", 0))
            self.redeemed = int(doc.get("redeemed", 0))

    @property
    def kid(self) -> str:
        return self.issuing_key.kid

    @property
    def public(self) -> B.PublicKey:
        return self.issuing_key.public

    def _persist(self) -> None:
        if not self.ledger_file:
            return
        d = os.path.dirname(os.path.abspath(self.ledger_file))
        fd, tmp = tempfile.mkstemp(dir=d, prefix=".pp-", suffix=".json")
        with os.fdopen(fd, "w") as f:
            json.dump({"kid": self.kid, "issued": self.issued, "redeemed": self.redeemed, "spent": sorted(self.spent)}, f)
        os.replace(tmp, self.ledger_file)

    # -- origin: challenge and redemption ----------------------------------

    def challenge(self, host: str) -> PP.TokenChallenge:
        """Static per-origin challenge (empty redemption context): tokens are pre-purchasable."""
        origin = self.origin_name or host
        return PP.TokenChallenge(PP.TOKEN_TYPE_BLIND_RSA, self.issuer_name or origin, b"", origin)

    def challenge_header(self, host: str) -> str:
        return PP.www_authenticate(self.challenge(host), self.public, max_age=86400 * 30)

    def redeem(self, authorization: str, host: str) -> tuple[bool, str]:
        try:
            token = PP.parse_authorization(authorization)
        except (ValueError, TypeError):
            return False, "malformed PrivateToken header"
        if token is None:
            return False, "no PrivateToken credential"
        if token.token_key_id not in self.keys:
            return False, "unknown token key"
        if not self.origin.verify(token, self.challenge(host)):
            return False, "invalid token or wrong challenge"
        nul = hashlib.sha256(b"nullifier" + token.nonce).hexdigest()
        with self._lock:
            if nul in self.spent:
                return False, "token already spent"
            self.spent.add(nul)
            self.redeemed += 1
            self._persist()
        return True, ""

    # -- issuer -------------------------------------------------------------

    def directory(self) -> dict:
        return PP.issuer_directory([k.public for k in self.keys.values()], "/token-request")

    def issue(self, body: bytes) -> bytes:
        req = PP.TokenRequest.deserialize(body)
        res = self.issuer.issue(req)
        with self._lock:
            self.issued += 1
            self._persist()
        return res.serialize()

    def issue_batch(self, body: bytes, expected: int | None = None) -> tuple[bytes, int]:
        reqs = PP.deserialize_batched_request(body)
        if expected is not None and len(reqs) != expected:
            raise ValueError(f"this batch endpoint issues exactly {expected} tokens; {len(reqs)} requested")
        if not 1 <= len(reqs) <= 100:
            raise ValueError("batches are 1-100 tokens")
        out: list[PP.TokenResponse | None] = []
        issued = 0
        for r in reqs:
            try:
                out.append(self.issuer.issue(r))
                issued += 1
            except ValueError:
                out.append(None)
        with self._lock:
            self.issued += issued
            self._persist()
        return PP.serialize_batched_response(out), issued

    def describe(self) -> dict:
        return {
            "standard": "RFC 9576/9577/9578, token type 0x0002 (Blind RSA, PSS)",
            "keys": [{"kid": k.kid, "issuing": k.issuing} for k in self.keys.values()],
            "kid": self.kid,
            "issuer_directory": PP.ISSUER_DIRECTORY_PATH,
            "issue": "POST /token-request (application/private-token-request or generic batch) or /token-request/batch/{10|25|100}",
            "redeem": "Authorization: PrivateToken token=...  (one token = one priced call)",
            "denominations": list(DENOMINATIONS),
            "price_per_token_usd": self.price_per_token_usd,
            "points_per_token": POINTS_PER_TOKEN,
            "pool": {"issued": self.issued, "redeemed": self.redeemed, "outstanding": self.issued - self.redeemed},
            "ephemeral_key": self.ephemeral,
        }


__all__ = ["PrivacyPassService", "TOKEN_IDENTITY", "DENOMINATIONS", "POINTS_PER_TOKEN", "generate_private_jwk"]
